ngraph/src/ngraph/pass/graph_rewrite.cpp

```diff
@@ -66,49 +66,51 @@ bool pass::GraphRewrite::run_on_function(shared_ptr<Function> f)
     bool rewritten = false;
     const size_t NUM_TRIES = 10;
     size_t tries = NUM_TRIES;
-    vector<MatchClosure> original_matchers{m_matchers};
-    // This check is very expensive and is only needed for experimental features, so we will hide
-    // it behind an environment variable for now. TODO: Find a less expensive way to handle this.
-    static bool s_rerun_dynamic_check = getenv_bool("NGRAPH_GRAPH_REWRITE_RERUN_DYNAMIC_CHECK");
-    bool is_dyn_func = s_rerun_dynamic_check && f->is_dynamic();
+    vector<std::shared_ptr<MatcherPass>> original_matchers{m_matchers};
     do
     {
         rewritten = false;
         // m_matchers may contain newly constructed matchers for matchers
         // that need multiple passes. See comments above.
-        vector<MatchClosure> matchers_to_run{m_matchers};
+        vector<std::shared_ptr<MatcherPass>> matchers_to_run{m_matchers};
         m_matchers.clear();
         for (auto node : f->get_ordered_ops())
         {
+            // Temporary keep this GraphRewrite property for backward compatibility
             if (m_enable_shape_inference)
             {
                 node->revalidate_and_infer_types();
             }
-            for (auto& closure : matchers_to_run)
+            for (auto& m_pass : matchers_to_run)
             {
-                if (is_dyn_func && closure.property[PassProperty::REQUIRE_STATIC_SHAPE])
+                if (m_pass->get_matcher_property().is_set(PassProperty::REQUIRE_STATIC_SHAPE) &&
+                    f->is_dynamic())
                 {
                     NGRAPH_DEBUG << "matcher callback requires static shape but the "
                                     "function is dynamic, skipping this "
                                     "optimization till the shapes are fully "
                                     "materialized";
                     continue;
                 }
-                if (closure.handler(node))
+
+                if (!m_has_default_callback)
+                {
+                    m_pass->set_callback(m_transformation_callback);
+                }
+
+                if (m_pass->apply_matcher(node))
                 {
                     rewritten = true;
-                    // If call back may change function's is_dynamic state, we need to
-                    // update the cached value.
-                    if (closure.property.is_set(PassProperty::CHANGE_DYNAMIC_STATE))
+                    if (m_pass->get_matcher_property().is_set(
+                            PassProperty::REQUIRE_SHAPE_INFERENCE))
                     {
-                        is_dyn_func = s_rerun_dynamic_check && f->is_dynamic();
+                        f->validate_nodes_and_infer_types();
```

Comment:
Is it too costly to evaluate entire function? Can we provide a way for MatcherPass to report about place in a function where the earliest (according to data-flow) changes happened to initiate shape inference at some point in the graph. Also, `Function::validate_and_infer_types` has too straight-forward implementation for our need here: it doesn't stop when shape and type inferred for the second time match completely the original ones. Can we think (probably in the following PR) about a way to avoid entire function re-evaluation?

Reply:
This was an experimental feature, we decided not to include it as is't not needed. Also PassProperty is going to be removed. Partial re-evaluatoin is a good feature, but lets first understand use cases and potential ways of implementation. I would keep  it as a separate feature/PR.